### PR TITLE
Fix #4936: Forbid (or warn on) redundant/illegal object flags

### LINKED ISSUE #4936: sealed final abstract case object Foo
Lots of modifiers are allowed on `object`s, but they do not seem to have any semantics.

In fact, you can have a
```scala
sealed final abstract case object Foo
```

But *sadly*
```scala
abstract object Foo {
  def x : String
}
```
doesn't compile, haha.

Perhaps those modifiers should be reported as errors?

### PR DESCRIPTION
**Pull request by Blaisorblade**
Forbid/warn on redundant/illegal object flags — fix #4936.

- Stop using final object in our sources.
- Give error for abstract and sealed objects, warning for final ones (as long as
 scala/bug#11094 is open).

### CODE REVIEW

**Comment by smarter** [comment]
could be done in the parser no?

**Comment by odersky** [comment]
@Blaisorblade Inactive for almost 5 months. Can you either finish the PR or close it and re-assign the issue to someone else? Thanks

File: tests/pos/i3471.scala [hunk 1]
```diff
@@ -1 +0,0 @@
-sealed object Fun
```
Comment (Blaisorblade) [inline tests/pos/i3471.scala]: This testcase was added in #3486 for #3471 to ensure the compiler didn't crash.
During the discussion, it was demanded that `final` be handled at the same time: https://github.com/lampepfl/dotty/pull/3486#discussion_r151471447.

**Comment by allanrenucci** [comment]
I think @odersky's [comment](https://github.com/lampepfl/dotty/pull/5525#issuecomment-447099112) applies to this PR

**Comment by Blaisorblade** [comment]
That comment does apply, and I guess it overrides contrary suggestions above.

**Comment by Blaisorblade** [comment]
FWIW, the current CI failure is spurious; while this is on hold, I'd like to get it working as far as possible.

**Comment by Blaisorblade** [comment]
@nicolasstucki I've done that change, so it could be merged, but @odersky in #5525 suggested centralizing flag checking. Assigning to you for decision.

**Comment by nicolasstucki** [comment]
We can centralize them later in #5525.

**Comment by odersky** [comment]
I am afraid this is the wrong fix. See our long discussion with Allan. Flag combinations should _not_ be tested in Parser, we discussed this and decided on it. Please let's revert these commits and do it from the ground up at the same time as #5525.

**Pull request merged**